File: android/measure/src/main/java/sh/measure/android/MsrAttachment.kt
package sh.measure.android

import sh.measure.android.events.Attachment
import sh.measure.android.events.AttachmentType
import java.io.File
import java.util.UUID

/**
 * An attachment which can be added to an event. Represents a file that provides additional context
 * or evidence for the reported issue.
 *
 * @property name The display name of the attachment, typically including the file extension.
 * @property bytes The bytes containing the attachment data.
 * @property type The type of attachment. Must be one of [AttachmentType].
 */
class MsrAttachment internal constructor(
    val name: String,
    val bytes: ByteArray?,
    val path: String? = null,
    val type: String,
) {
    constructor(
        name: String,
        path: String?,
        type: String,
    ) : this(name, null, path, type)

    init {
        require(type in AttachmentType.VALID_TYPES) { "Invalid attachment type: $type" }
    }
}

internal fun MsrAttachment.toEventAttachment(attachmentType: String): Attachment {
    val id = UUID.randomUUID().toString()
    val size = bytes?.size?.toLong() ?: path?.let {
        try {
            File(it).length()
        } catch (_: Exception) {
            0L
        }
    } ?: 0L
    return Attachment(
        id = id,
        name = name,
        type = attachmentType,
        size = size,
        bytes = bytes,
        path = path,
    )
}

File: ios/Sources/MeasureSDK/Swift/MsrAttachment.swift
//
//  MsrAttachment.swift
//  Measure
//
//  Created by Adwin Ross on 08/05/25.
//

import Foundation

/// An attachment which can be added to an event. Represents a file that provides additional context
/// or evidence for the reported issue.
/// - Parameters:
///   - name: The display name of the attachment, typically including the file extension.
///   - bytes: The raw data of the attachment.
///   - type: The type of attachment, typically one of `AttachmentType`.
public class MsrAttachment: NSObject {
    let name: String
    let bytes: Data
    let type: AttachmentType

    init(name: String, bytes: Data, type: AttachmentType) {
        self.name = name
        self.bytes = bytes
        self.type = type
    }
}

extension MsrAttachment {
    func toEventAttachment(id: String) -> Attachment {
        return Attachment(name: self.name,
                          type: self.type,
                          size: Int64(self.bytes.count),
                          id: id,
                          bytes: self.bytes,
                          path: nil)
    }
}
